### PR TITLE
Support X-Forwarded-For when it comes from a trusted proxy

### PR DESCRIPTION
**Pull request by SeanBurford**
I find this useful for running routedns behind nginx.

### CODE REVIEW

**Comment by cbuijs** [comment]
Maybe use same approach as with EDNS (like for subnet caching), and support
for CIDR addresses to designated trusted proxies/subnets?

Not sure about the "proxy" naming convention as it is more http related
then DNS but from the usage/application it makes sense.

**Comment by SeanBurford** [comment]
proxy-subnet is a good idea, however I would prefer to push that back to a future PR because the config parsing will add complexity and:

- Config (trusted-proxy-subnet4, trusted-proxy-subnet6) can be added in a backwards compatible way (default to a subnet of all ones, which is the current behaviour).
- Implementation is straightforward (Change ip.Equal to subnet.Contains).

File: cmd/routedns/main.go [hunk 1]
```diff
@@ -244,7 +244,17 @@ func start(opt options, args []string) error {
...
-			ln, err := rdns.NewDoHListener(id, l.Address, rdns.DoHListenerOptions{TLSConfig: tlsConfig, ListenOptions: opt, Transport: l.Transport}, resolver)
+			proxyAddr := net.ParseIP(l.DoH.ProxyAddr)
+			if l.DoH.ProxyAddr != "" && proxyAddr == nil {
+				return fmt.Errorf("listener '%s' proxy-address '%s' is not an IPv4 or IPv6 address", id, l.DoH.ProxyAddr)
+			}
+			opt := rdns.DoHListenerOptions{
+				TLSConfig:     tlsConfig,
+				ListenOptions: opt,
+				Transport:     l.Transport,
+				ProxyAddr:     &proxyAddr,
```
Comment (folbricht) [inline cmd/routedns/main.go:255]: Do we really need to pass a pointer here? net.IP is already a slice and I believe it would work without it

Comment (SeanBurford) [reply cmd/routedns/main.go]: Thanks, I've removed the pointer.

**Comment by SeanBurford** [comment]
> Not sure about the "proxy" naming convention as it is more http related then DNS but from the usage/application it makes sense.

I've moved to HttpProxy to make the intention clearer.  Since these are listeners the HTTPProxy must be incoming rather than outbound.

**Comment by folbricht** [comment]
Thanks for the contribution!

**Pull request merged**